### PR TITLE
Improve piston handling efficiency

### PR DESCRIPTION
**Pull request by Jikoo**
This is a complete overhaul of piston handling, so it's a bit of a slog to read through. I did my best to keep it well-commented, but you may be better off reading the end result instead of the diff.

* Added configuration option `GriefPrevention.PistonMovement`
  * `EVERYWHERE`: Pistons can be used everywhere. All blocks moved are checked to ensure that they do not enter into a claim that is not owned by the owner of the piston's claim.
  * `EVERYWHERE_SIMPLE`: Pistons can be used everywhere. Instead of checking every block moved, the zone surrounding all moved blocks is checked. As long as no part of the zone overlaps a claim with a different owner, pistons are allowed to move freely. This means that certain non-threatening movements will be blocked, but it handles more complex structures much faster.
Example of a valid but blocked piston:
image
  * `CLAIMS_ONLY` (default): Pistons may only be used inside claims. If blocks outside of the claim that the piston is in are affected (even if the owner is the same!) the piston movement is blocked. This is the only mode that completely prevents pushing blocks out of claims.
  * `IGNORED`: Pistons are not checked. Griefers can use flying machines to destroy or disrupt your builds.
* Removed configuration options `GriefPrevention.LimitPistonsToLandClaims` and `GriefPrevention.CheckPistonMovement`
  * Settings are migrated to new equivalents

For server owners: if you have already turned off CheckPistonMovement I would appreciate it if you'd give one of the reworked types a shot. My personal favorite is `EVERYWHERE_SIMPLE`, but `CLAIMS_ONLY` is the most performant.

### CODE REVIEW

**Comment by Jikoo** [comment]
I did replace the old metrics chart with one for the new modes, not sure if that's actually desirable for your viewing pleasure. Pretty sure that does require a change on the bStats end.

**Comment by RoboMWM** [comment]
> Pretty sure that does require a change on the bStats end.

if it's just reporting different strings, then no change needed. Simple pie should be sufficient, which seems to be the only chart that's ever used anywhere lol.

gonna try to work on the gp docs repo so we can get this documented.

**Comment by Jikoo** [comment]
> if it's just reporting different strings, then no change needed. Simple pie should be sufficient, which seems to be the only chart that's ever used anywhere lol.

I did move it to a new pie so that it would be less confusing (figured true/false mixed in wouldn't play well), would you like to to stay in the same place? It is a simple pie, yeah.

**Comment by RoboMWM** [comment]
Oh I see what you mean, it'll conflict with existing data. Hmm... I'd keep it the same id anyways, the percentages won't add up to 100% until everyone has moved to this version but we'd still be able to see the relative portions regardless which is what we care about.

File: src/main/java/me/ryanhamshire/GriefPrevention/PistonMode.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package me.ryanhamshire.GriefPrevention;
+
+public enum PistonMode {
+
+    EVERYWHERE,
+    EVERYWHERE_FAST,
```
Comment (RoboMWM) [inline src/main/java/me/ryanhamshire/GriefPrevention/PistonMode.java]: maybe EVERYWHERE_SIMPLE

**Review by RoboMWM — DISMISSED** [review]
pistons are annoying.

**Comment by Jikoo** [comment]
Sorry about the mess, no power makes this a little difficult.

**Review by RoboMWM — APPROVED** [review]
Something has to remind me to put this in the documentation, so I'll probably hold off on pulling this and instead have it ready for the next release

**Comment by jacob1** [comment]
There is one issue with this PR, which is that it doesn't work in admin claims.
The new code is `pistonClaim.ownerID.equals(claim.ownerID)`, but ownerID is [null for admin claims](https://github.com/TechFortress/GriefPrevention/blob/master/src/main/java/me/ryanhamshire/GriefPrevention/Claim.java#L57). So a NullPointerException is thrown.

Players on my server were coincidentally looking at this exact lag issue today, after we noticed server tps depended on whether a certain player was online. I was going to open a PR changing it to use ownerID, then I saw this open PR. I guess it got merged before I could comment about the admin claim issue.

Here is the stacktrace which now happens when you use a piston in an admin claim:
```[20:59:31 ERROR]: Could not pass event BlockPistonExtendEvent to GriefPrevention v16.15.0-16611fb
org.bukkit.event.EventException: null
	at org.bukkit.plugin.java.JavaPluginLoader$1.execute(JavaPluginLoader.java:319) ~[spigot-1.16.3.jar:git-Spigot-3a70bd9-bf3d720]
	at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70) ~[spigot-1.16.3.jar:git-Spigot-3a70bd9-bf3d720]
	at org.bukkit.plugin.SimplePluginManager.fireEvent(SimplePluginManager.java:589) ~[spigot-1.16.3.jar:git-Spigot-3a70bd9-bf3d720]
	at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:576) ~[spigot-1.16.3.jar:git-Spigot-3a70bd9-bf3d720]
	at net.minecraft.server.v1_16_R2.BlockPiston.a(BlockPiston.java:330) ~[spigot-1.16.3.jar:git-Spigot-3a70bd9-bf3d720]
	at net.minecraft.server.v1_16_R2.BlockPiston.a(BlockPiston.java:183) ~[spigot-1.16.3.jar:git-Spigot-3a70bd9-bf3d720]
	at net.minecraft.server.v1_16_R2.BlockBase$BlockData.a(BlockBase.java:493) ~[spigot-1.16.3.jar:git-Spigot-3a70bd9-bf3d720]
	at net.minecraft.server.v1_16_R2.WorldServer.a(WorldServer.java:1293) ~[spigot-1.16.3.jar:git-Spigot-3a70bd9-bf3d720]
	at net.minecraft.server.v1_16_R2.WorldServer.aj(WorldServer.java:1283) ~[spigot-1.16.3.jar:git-Spigot-3a70bd9-bf3d720]
	at net.minecraft.server.v1_16_R2.WorldServer.doTick(WorldServer.java:342) ~[spigot-1.16.3.jar:git-Spigot-3a70bd9-bf3d720]
	at net.minecraft.server.v1_16_R2.MinecraftServer.b(MinecraftServer.java:1104) ~[spigot-1.16.3.jar:git-Spigot-3a70bd9-bf3d720]
	at net.minecraft.server.v1_16_R2.DedicatedServer.b(DedicatedServer.java:352) ~[spigot-1.16.3.jar:git-Spigot-3a70bd9-bf3d720]
	at net.minecraft.server.v1_16_R2.MinecraftServer.a(MinecraftServer.java:1007) ~[spigot-1.16.3.jar:git-Spigot-3a70bd9-bf3d720]
	at net.minecraft.server.v1_16_R2.MinecraftServer.w(MinecraftServer.java:846) ~[spigot-1.16.3.jar:git-Spigot-3a70bd9-bf3d720]
	at net.minecraft.server.v1_16_R2.MinecraftServer.lambda$0(MinecraftServer.java:164) ~[spigot-1.16.3.jar:git-Spigot-3a70bd9-bf3d720]
	at java.lang.Thread.run(Thread.java:835) [?:?]
Caused by: java.lang.NullPointerException
	at me.ryanhamshire.GriefPrevention.BlockEventHandler.onPistonEvent(BlockEventHandler.java:519) ~[?:?]
	at me.ryanhamshire.GriefPrevention.BlockEventHandler.onBlockPistonExtend(BlockEventHandler.java:483) ~[?:?]
	at jdk.internal.reflect.GeneratedMethodAccessor9.invoke(Unknown Source) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:567) ~[?:?]
	at org.bukkit.plugin.java.JavaPluginLoader$1.execute(JavaPluginLoader.java:315) ~[spigot-1.16.3.jar:git-Spigot-3a70bd9-bf3d720]
	... 15 more```

**Comment by RoboMWM** [comment]
Thanks for testing! And ya, had some time to push out some updates and go through PRs.

File: src/main/java/me/ryanhamshire/GriefPrevention/GriefPrevention.java [hunk 2]
```diff
@@ -625,12 +624,11 @@ else if (world.getEnvironment() == Environment.NORMAL)
...
         this.config_blockSkyTrees = config.getBoolean("GriefPrevention.LimitSkyTrees", true);
         this.config_limitTreeGrowth = config.getBoolean("GriefPrevention.LimitTreeGrowth", false);
-        this.config_checkPistonMovement = config.getBoolean("GriefPrevention.CheckPistonMovement", true);
-        this.config_pistonsInClaimsOnly = config.getBoolean("GriefPrevention.LimitPistonsToLandClaims", true);
-        if (!this.config_checkPistonMovement && this.config_pistonsInClaimsOnly) {
-            AddLogEntry("Error: You have enabled LimitPistonsToLandClaims, but CheckPistonMovement is off!");
-            this.config_pistonsInClaimsOnly = false;
-        }
+        this.config_pistonMovement = PistonMode.of(config.getString("GriefPrevention.PistonMovement", "CLAIMS_ONLY"));
+        if (config.isBoolean("GriefPrevention.LimitPistonsToLandClaims") && !config.isBoolean("GriefPrevention.LimitPistonsToLandClaims"))
```
Comment (Bobcat00) [inline src/main/java/me/ryanhamshire/GriefPrevention/GriefPrevention.java:628]: Looks like a typo. You have isBoolean twice in line 628.

Comment (Jikoo) [reply src/main/java/me/ryanhamshire/GriefPrevention/GriefPrevention.java:628]: Good catch, thanks!

**Pull request merged**